File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/ShareActivity.java
package ch.rmy.android.http_shortcuts;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import ch.rmy.android.http_shortcuts.realm.Controller;
import ch.rmy.android.http_shortcuts.realm.models.Shortcut;
import ch.rmy.android.http_shortcuts.realm.models.Variable;
import ch.rmy.android.http_shortcuts.utils.IntentUtil;
import ch.rmy.android.http_shortcuts.utils.MenuDialogBuilder;
import ch.rmy.android.http_shortcuts.variables.VariableResolver;

public class ShareActivity extends BaseActivity {

    private static final String TYPE_TEXT = "text/plain";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        String type = getIntent().getType();
        if (!TYPE_TEXT.equals(type)) {
            finishWithoutAnimation();
            return;
        }
        String text = getIntent().getStringExtra(Intent.EXTRA_TEXT);
        if (text == null) {
            showInstructions(R.string.error_sharing_feature_not_available_yet);
            return;
        }

        Controller controller = destroyer.own(new Controller(getContext()));
        Set<String> variables = getTargetableVariables(controller);
        List<Shortcut> shortcuts = getTargetableShortcuts(controller, variables);

        HashMap<String, String> variableValues = new HashMap<>();
        for (String variable : variables) {
            variableValues.put(variable, text);
        }

        if (shortcuts.size() == 1) {
            executeShortcut(shortcuts.get(0), variableValues);
            finishWithoutAnimation();
        } else if (shortcuts.isEmpty()) {
            showInstructions(R.string.error_not_suitable_shortcuts);
        } else {
            showShortcutSelection(shortcuts, variableValues);
        }

        // TODO:
        // - Write documentation
        // - Variable input auto complete
        // - Translate
    }

    private Set<String> getTargetableVariables(Controller controller) {
        Set<String> targetableVariables = new HashSet<>();
        List<Variable> variables = controller.getVariables();
        for (Variable variable : variables) {
            if (variable.isShareText()) {
                targetableVariables.add(variable.getKey());
            }
        }
        return targetableVariables;
    }

    private List<Shortcut> getTargetableShortcuts(Controller controller, Set<String> variableKeys) {
        List<Shortcut> shortcuts = new ArrayList<>();
        for (Shortcut shortcut : controller.getShortcuts()) {
            if (hasShareVariable(shortcut, variableKeys)) {
                shortcuts.add(shortcut);
            }
        }
        return shortcuts;
    }

    private boolean hasShareVariable(Shortcut shortcut, Set<String> variableKeys) {
        Set<String> variableKeysInShortcut = VariableResolver.extractVariableKeys(shortcut);
        for (String key : variableKeys) {
            if (variableKeysInShortcut.contains(key)) {
                return true;
            }
        }
        return false;
    }

    private void executeShortcut(Shortcut shortcut, HashMap<String, String> variableValues) {
        Intent intent = IntentUtil.createIntent(getContext(), shortcut.getId(), variableValues);
        startActivity(intent);
    }

    private void showInstructions(@StringRes int text) {
        new MaterialDialog.Builder(getContext())
                .content(text)
                .dismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        finishWithoutAnimation();
                    }
                })
                .positiveText(R.string.button_ok)
                .show();
    }

    private void showShortcutSelection(List<Shortcut> shortcuts, final HashMap<String, String> variableValues) {
        MenuDialogBuilder builder = new MenuDialogBuilder(getContext());
        for (final Shortcut shortcut : shortcuts) {
            builder.item(shortcut.getName(), new MenuDialogBuilder.Action() {
                @Override
                public void execute() {
                    executeShortcut(shortcut, variableValues);
                }
            });
        }
        builder.dismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                finishWithoutAnimation();
            }
        }).show();
    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/activities/ShareActivity.kt
package ch.rmy.android.http_shortcuts.activities

import android.content.Intent
import android.os.Bundle
import android.support.annotation.StringRes
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.dialogs.MenuDialogBuilder
import ch.rmy.android.http_shortcuts.realm.Controller
import ch.rmy.android.http_shortcuts.realm.models.Shortcut
import ch.rmy.android.http_shortcuts.utils.mapFor
import ch.rmy.android.http_shortcuts.utils.showIfPossible
import ch.rmy.android.http_shortcuts.variables.VariableResolver
import com.afollestad.materialdialogs.MaterialDialog

class ShareActivity : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val type = intent.type
        if (type != TYPE_TEXT) {
            finishWithoutAnimation()
            return
        }
        val text = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (text == null) {
            showInstructions(R.string.error_sharing_feature_not_available_yet)
            return
        }

        val controller = destroyer.own(Controller())
        val variables = getTargetableVariables(controller)
        val shortcuts = getTargetableShortcuts(controller, variables)

        val variableValues = variables.associate { variable ->
            variable to text
        }

        when (shortcuts.size) {
            0 -> showInstructions(R.string.error_not_suitable_shortcuts)
            1 -> {
                executeShortcut(shortcuts[0], variableValues)
                finishWithoutAnimation()
            }
            else -> showShortcutSelection(shortcuts, variableValues)
        }
    }

    private fun getTargetableVariables(controller: Controller) =
            controller
                    .getVariables()
                    .filter { it.isShareText }
                    .map { it.key }
                    .toSet()

    private fun getTargetableShortcuts(controller: Controller, variableKeys: Set<String>): List<Shortcut> =
            controller
                    .getShortcuts()
                    .filter { hasShareVariable(it, variableKeys) }

    private fun hasShareVariable(shortcut: Shortcut, variableKeys: Set<String>): Boolean {
        val variableKeysInShortcut = VariableResolver.extractVariableKeys(shortcut)
        return variableKeys.any { variableKeysInShortcut.contains(it) }
    }

    private fun executeShortcut(shortcut: Shortcut, variableValues: Map<String, String>) {
        val intent = ExecuteActivity.IntentBuilder(context, shortcut.id)
                .variableValues(variableValues)
                .build()
        startActivity(intent)
    }

    private fun showInstructions(@StringRes text: Int) {
        MaterialDialog.Builder(context)
                .content(text)
                .dismissListener { finishWithoutAnimation() }
                .positiveText(R.string.dialog_ok)
                .showIfPossible()
    }

    private fun showShortcutSelection(shortcuts: List<Shortcut>, variableValues: Map<String, String>) {
        MenuDialogBuilder(context)
                .mapFor(shortcuts) { builder, shortcut ->
                    builder.item(shortcut.name) {
                        executeShortcut(shortcut, variableValues)
                    }
                }
                .dismissListener { finishWithoutAnimation() }
                .showIfPossible()
    }

    companion object {

        private const val TYPE_TEXT = "text/plain"

    }

}
